Features/UI grouping

Meertman opened the pull request:
<!--  Thanks for sending a pull request!
-->

**What this PR does / why we need it**:
This PR allows the end-user to group his health check endpoints into either custom defined groups, or in case the end-user is using the kubernetes discovery it will group the health checks using the namespace in which the pods that execute the healthchecks are hosted.

**Which issue(s) this PR fixes**: #788, #690, groundwork for #809 

Please reference the issue this PR will close: #788 #690 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [X] Code compiles correctly
- [ ] Created/updated tests
- [ ] Unit tests passing
- [X] End-to-end tests passing
- [ ] Extended the documentation
- [ ] Provided sample for the feature

Meertman commented:
Credits where credits are due, the UI work has been done by @GeniaT

BoSoeborgPetersen commented:
I love the idea of adding groupings to the HealthCheckUI, but I had to add "dom.iterable" to tsconfig.json -> "compilerOptions" -> "lib" in HealthChecksUI project.

And still when I run the HealthChecks.UI.Branding sample I see this:
![image](https://user-images.githubusercontent.com/11646592/133753749-6dd09f69-f5d8-4590-847c-c870a6cdaf79.png)
And this when expanded (which can only be done with the expand all button:
![image](https://user-images.githubusercontent.com/11646592/133753897-986a04de-e625-4548-ae67-d195b5034972.png)

I would assume that groups are optional and that all health checks without a group would end up in a default group?

Meertman commented:
@BoSoeborgPetersen, sorry, I did indeed forget to check in that change to the tsconfig.json file. (I added it to the PR)

So yes, health checks without a group end up in the default group. Which currently has no name :)

BoSoeborgPetersen commented:
@Meertman, from what I can tell, the backend works as expected, but there are some frontend issues:

There is a problem with the expand/collapse button at the group level:
 - Default group: it does nothing
 - Other groups: it works, but it also expands the health checks in the group
I think the problem is that the typescript does not react well to the default group having null as the name
 
I would also suggest that:
 - Groups have a different background color than health checks
 - If only the default group exists, then just display the health checks without the group level (no API changes, just UI), this would ensure that current users don't see any changes because of this feature, in the UI, and they can then choose to use this feature. But maybe this should also be part of the API, so the data structure is the same for people not using this feature, otherwise we could break peoples use of the API from somewhere other than the UI.

I also found a possible mistake on line 58 in IEndpointRouteBuilderExtensions.cs, it should probably be AddAsync, instead of RemoveAsync.

I have added the background color for groups locally, and will look further at the other parts, but I am limited by my javascript/typescript/frontend skills, and I am also not sure where to push my changes, possibly a fork of your fork.

smbecker commented on build/docker-images/HealthChecks.UI.Image/Extensions/IEndpointRouteBuilderExtensions.cs:
```diff
@@ -54,15 +55,15 @@ public static class IEndpointRouteBuilderExtensions
                         if (type == PushServiceKeys.ServiceAdded)
                         {
-                            await pushService.AddAsync(name, uri);
+                            await pushService.RemoveAsync(name, group);
```
Shouldn't this be `AddAsync`?

smbecker commented:
It would be nice to be able to include the group name in [webhook notifications](https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks/blob/50594f0643515e83f9bf6ce864f9594bf980b865/src/HealthChecks.UI/Core/Notifications/WebHookFailureNotifier.cs#L78).

Meertman commented:
I am also limited by my knowledge of React, so if someone could take a look at the UI part, this might be merged as well?

Meertman commented:
> @Meertman, from what I can tell, the backend works as expected, but there are some frontend issues:
> 
> There is a problem with the expand/collapse button at the group level:
> 
> * Default group: it does nothing
> * Other groups: it works, but it also expands the health checks in the group
>   I think the problem is that the typescript does not react well to the default group having null as the name
> 
> I would also suggest that:
> 
> * Groups have a different background color than health checks
> * If only the default group exists, then just display the health checks without the group level (no API changes, just UI), this would ensure that current users don't see any changes because of this feature, in the UI, and they can then choose to use this feature. But maybe this should also be part of the API, so the data structure is the same for people not using this feature, otherwise we could break peoples use of the API from somewhere other than the UI.
> 
> I also found a possible mistake on line 58 in IEndpointRouteBuilderExtensions.cs, it should probably be AddAsync, instead of RemoveAsync.
> 
> I have added the background color for groups locally, and will look further at the other parts, but I am limited by my javascript/typescript/frontend skills, and I am also not sure where to push my changes, possibly a fork of your fork.

@BoSoeborgPetersen for your information, @Tom-De-Backer has implemented these changes and I've added them to the pull request.

@sungam3r, can we get these changes merged with the main repo?

sungam3r commented:
@Meertman Sorry but I'm not familiar and I don't use such areas of this repo as docker/JS/UI so better to ask maintainers. I just help to keep things moving in HC packages, CI and around some other stuff.

TrueBurn commented:
Is it possible to have the group come from an annotation on the service (the annotation should be configurable) then maybe with the k8s namespace as a fallback if there is no group set via the service annotation?
